src/validate.js

```diff
@@ -102,23 +107,214 @@ class ApplyRules {
+      fieldsToTest = data[field];
+    } else if (typeof data[field] === 'object' && data[field] !== null) {
+      fieldsToTest.push(data[field]);
+      isSingleObject = true;
+    }
+
+    let index = 0;
+    for (const fieldValue of fieldsToTest) {
+      if (typeof fieldValue === 'object' && fieldValue !== null && !(fieldValue instanceof Array)) {
+        const subModelType = PropertyHelper.getObjectField(fieldValue, '@type', nodeToTest.options.version);
+        const currentFieldName = field;
+        let currentFieldIndex;
+        if (!isSingleObject) {
+          currentFieldIndex = index;
+        }
+
+        // Check this is not a value object
+        if (
+          typeof fieldValue['@value'] === 'undefined'
+        ) {
+          let modelObj = this.loadModel(
+            subModelType,
+            nodeToTest.options.version,
+          );
+          if (!modelObj.hasSpecification) {
+            // Try loading from the parent model type if we can
+            if (
+              typeof nodeToTest.model.fields[field] !== 'undefined'
+              && typeof nodeToTest.model.fields[field].model !== 'undefined'
+            ) {
+              const altSubModelType = nodeToTest.model.fields[field].model.replace(/^(ArrayOf)?#/, '');
+              if (
+                typeof nodeToTest.model.fields[field].requiredType === 'undefined'
+                && typeof nodeToTest.model.fields[field].alternativeTypes === 'undefined'
+                && typeof nodeToTest.model.fields[field].alternativeModels === 'undefined'
+              ) {
+                modelObj = this.loadModel(
+                  altSubModelType,
+                  nodeToTest.options.version,
+                );
+              }
+            }
+          }
+
+          const newNodeToTest = new ModelNode(
+            currentFieldName,
+            fieldValue,
+            nodeToTest,
+            modelObj,
+            nodeToTest.options,
+            currentFieldIndex,
+          );
+
+          const subModelErrors = await this.applyModelRulesAsync(
+            rules,
+            newNodeToTest,
+          );
+          errors = errors.concat(subModelErrors);
+        }
+      }
+      index += 1;
+    }
+    return errors;
+  }
+
+  static async applyModelRulesAsync(rules, nodeToTest) {
+    let errors = [];
+    for (const rule of rules) {
+      const newErrors = await rule.validateAsync(nodeToTest);
+      // Apply whole-model rule, and field-specific rules
+      errors = errors.concat(newErrors);
+    }
+    for (const field in nodeToTest.value) {
+      if (Object.prototype.hasOwnProperty.call(nodeToTest.value, field)) {
+        // If this field is itself a model, apply that model's rules to it
+        const subModelErrors = await ApplyRules.applySubModelRulesAsync(rules, nodeToTest, field);
+        errors = errors.concat(subModelErrors);
+      }
+    }
+    return errors;
+  }
+}
+
+async function validateAsync(value, options) {
```

Comment:
This is almost identical to the previous `validate` method except uses await where appropriate